test_quality_of_service/test/test_liveliness.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+#include <chrono>
+#include <exception>
+#include <iostream>
+#include <memory>
+#include <stack>
+#include <string>
+#include <tuple>
+
+#include "gtest/gtest.h"
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp/executors/single_threaded_executor.hpp"
+
+#include "std_msgs/msg/string.hpp"
+
+#include "test_quality_of_service/qos_utilities.hpp"
+
+using namespace std::chrono_literals;
+
+/// Liveliness setup
+class LivelinessSetup : public ::testing::Test
+{
+protected:
+  void SetUp() override
+  {
+    rclcpp::init(0, nullptr);
+    executor = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+  }
+  void TearDown() override
+  {
+    executor->cancel();
+    executor.reset();
+    rclcpp::shutdown();
+  }
+
+  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> executor;
+};
+
+/// Test Automatic Liveliness with a single publishing node and single subscriber node
+TEST_F(LivelinessSetup, test_automatic_liveliness_changed) {
+  const std::chrono::milliseconds max_test_length = 8s;
+  const std::chrono::milliseconds kill_publisher_after = 2s;
+  int number_of_published_messages = 0;
+
+  // used for lambda capture
+  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec = executor;
+  int total_number_of_liveliness_events = 0;
+
+  rmw_qos_profile_t qos_profile = rmw_qos_profile_default;
+  // define liveliness policy
+  qos_profile.liveliness = RMW_QOS_POLICY_LIVELINESS_AUTOMATIC;
+
+  std::tuple<size_t, size_t> liveliness_lease_tuple = chrono_milliseconds_to_size_t(1000ms);
```

Comment:
This should be implemented such that you can just write `qos_profile.liveliness_lease_duration = chrono_duration_to_rmw_time(lease_duration)`